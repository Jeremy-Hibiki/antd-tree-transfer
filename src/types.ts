import type { TransferProps, TreeDataNode, TreeProps } from 'antd';
import type { ReactNode } from 'react';

type Simplify<T> = { [P in keyof T]: T[P] } & {};

export type TreeTransferItem = Simplify<
  {
    children?: TreeTransferItem[];
    title: string;
  } & Omit<TreeDataNode, 'children' | 'title'>
>;

export type TreeTransferProps<RecordType extends TreeTransferItem = TreeTransferItem> = Simplify<
  {
    dataSource: RecordType[];
    treeProps?:
      | {
          leftTree?: TreeProps;
          rightTree?: TreeProps;
        }
      | TreeProps;
    loadingRenderer?: ReactNode;
  } & TransferProps<RecordType>
>;
