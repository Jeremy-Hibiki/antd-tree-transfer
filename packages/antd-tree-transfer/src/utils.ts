import type { TreeDataNode } from 'antd';
import type { Key } from 'react';

export const generateLeftTree = (treeNodes: TreeDataNode[] = [], targetKeys: Key[] = []): TreeDataNode[] => {
  return treeNodes.map((node) => {
    const newNode = { ...node };
    newNode.disabled = targetKeys.includes(node.key);
    if (newNode.children) {
      newNode.children = generateLeftTree(newNode.children, targetKeys);
      if (!newNode.disabled) {
        newNode.disabled = newNode.children.every((child) => child.disabled);
      }
    }
    return newNode;
  });
};

export const generateRightTree = (treeNodes: TreeDataNode[] = [], targetKeys: Key[] = []): TreeDataNode[] => {
  return treeNodes
    .map(({ children, ...props }) => {
      return {
        ...props,
        _keepNodeFlag: targetKeys.includes(props.key),
        children: generateRightTree(children, targetKeys),
      };
    })
    .filter(
      (node) =>
        node._keepNodeFlag ||
        (node.children.length && node.children.some((child) => (child as typeof node)._keepNodeFlag)),
    );
};

export const searchTreeFilter = (
  treeNodes: TreeDataNode[] = [],
  filteredItems: TreeDataNode[] = [],
): TreeDataNode[] => {
  const filteredKeys = filteredItems.map((item) => item.key);
  // filter out the nodes in the filteredItems and its ancestors
  // don't mutate the original treeNodes
  const result: TreeDataNode[] = [];
  treeNodes.forEach((node) => {
    const newNode = { ...node };
    if (newNode.children) {
      newNode.children = searchTreeFilter(newNode.children, filteredItems);
      if (newNode.children.length !== 0) {
        result.push(newNode);
      }
    } else if (filteredKeys.includes(newNode.key)) {
      result.push(newNode);
    }
  });
  return result;
};
