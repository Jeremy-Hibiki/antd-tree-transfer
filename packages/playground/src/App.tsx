import type { TreeDataNode } from 'antd';

import './App.css';

import TreeTransfer from 'antd-tree-transfer';
import { type Key, useState } from 'react';

enum TreeNodeKey {
  PARENT_1_WO_CHILDREN = 'parent-1-without-children',
  PARENT_2_WITH_ONE_CHILD = 'parent-2-with-one-child',
  PARENT_2_CHILD_1 = 'parent-2-child-1',
  PARENT_3_WITH_MULTIPLE_CHILDREN = 'parent-3-with-multiple-children',
  PARENT_3_CHILD_1 = 'parent-3-child-1',
  PARENT_3_CHILD_2 = 'parent-3-child-2',
  PARENT_3_CHILD_3 = 'parent-3-child-3',
}

const TREE_NODES = [
  {
    key: TreeNodeKey.PARENT_1_WO_CHILDREN,
    title: TreeNodeKey.PARENT_1_WO_CHILDREN,
  },
  {
    key: TreeNodeKey.PARENT_2_WITH_ONE_CHILD,
    title: TreeNodeKey.PARENT_2_WITH_ONE_CHILD,
    children: [
      {
        key: TreeNodeKey.PARENT_2_CHILD_1,
        title: TreeNodeKey.PARENT_2_CHILD_1,
      },
    ],
  },
  {
    key: TreeNodeKey.PARENT_3_WITH_MULTIPLE_CHILDREN,
    title: TreeNodeKey.PARENT_3_WITH_MULTIPLE_CHILDREN,
    children: [
      {
        key: TreeNodeKey.PARENT_3_CHILD_1,
        title: TreeNodeKey.PARENT_3_CHILD_1,
      },
      {
        key: TreeNodeKey.PARENT_3_CHILD_2,
        title: TreeNodeKey.PARENT_3_CHILD_2,
      },
      {
        key: TreeNodeKey.PARENT_3_CHILD_3,
        title: TreeNodeKey.PARENT_3_CHILD_3,
      },
    ],
  },
] as const satisfies TreeDataNode[];

// eslint-disable-next-line ts/explicit-function-return-type
function App() {
  const [targetKeys, setTargetKeys] = useState<Key[]>([]);

  return (
    <TreeTransfer
      dataSource={TREE_NODES}
      targetKeys={targetKeys}
      onChange={(keys) => {
        setTargetKeys(keys);
      }}
    />
  );
}

export default App;
