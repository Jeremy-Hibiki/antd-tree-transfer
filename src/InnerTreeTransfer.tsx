import type { Key } from 'react';

import { theme, Transfer, Tree } from 'antd';
import { useMemo } from 'react';

import type { TreeTransferItem, TreeTransferProps } from './types';

import { generateLeftTree, generateRightTree, searchTreeFilter } from './utils';

const InnerTreeTransfer = <RecordType extends TreeTransferItem = TreeTransferItem>({
  dataSource,
  loadingRenderer,
  targetKeys,
  treeProps,
  ...restProps
}: TreeTransferProps<RecordType>): ReturnType<typeof Transfer<RecordType>> => {
  const { token } = theme.useToken();

  const transferDataSource = useMemo(() => {
    const result: RecordType[] = [];
    function flatten(list: RecordType[] = []): void {
      list.forEach((item) => {
        item.isLeaf = !item.children || item.children.length === 0;
        result.push(item);
        flatten(item.children as RecordType[]);
      });
    }
    flatten(dataSource);
    return result;
  }, [dataSource]);

  const [leftTreeProps, rightTreeProps] = useMemo(() => {
    if (!treeProps) {
      return [{}, {}];
    }
    if ('leftTree' in treeProps || 'rightTree' in treeProps) {
      return [treeProps.leftTree ?? {}, treeProps.rightTree ?? {}];
    }
    return [treeProps, treeProps];
  }, [treeProps]);

  return (
    <Transfer<RecordType>
      {...restProps}
      className="tree-transfer"
      dataSource={transferDataSource}
      targetKeys={targetKeys}
      showSelectAll
    >
      {({ direction, filteredItems, onItemSelectAll, selectedKeys }) => {
        if (transferDataSource.length === 0) {
          return loadingRenderer;
        }

        const items = searchTreeFilter(dataSource, filteredItems);

        if (direction === 'left') {
          const checkedKeys = [...selectedKeys, ...(targetKeys ?? [])];
          const treeData = generateLeftTree(items, targetKeys);
          return (
            <div style={{ padding: token.paddingXS }}>
              <Tree
                checkedKeys={checkedKeys}
                treeData={treeData}
                blockNode
                checkable
                showLine
                onCheck={(_, { checkedNodes }) => {
                  onItemSelectAll(
                    [...new Set([...(targetKeys ?? []), ...checkedNodes.map((node) => node.key)])],
                    'replace',
                  );
                }}
                {...leftTreeProps}
              />
            </div>
          );
        } else {
          const treeData = generateRightTree(items, targetKeys);
          return (
            <div style={{ padding: token.paddingXS }}>
              <Tree
                checkedKeys={selectedKeys}
                treeData={treeData}
                blockNode
                checkable
                defaultExpandAll
                showLine
                onCheck={(checked) => {
                  onItemSelectAll(checked as Key[], 'replace');
                }}
                {...rightTreeProps}
              />
            </div>
          );
        }
      }}
    </Transfer>
  );
};

export default InnerTreeTransfer;
