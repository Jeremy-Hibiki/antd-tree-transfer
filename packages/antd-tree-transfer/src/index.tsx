import type { FC } from 'react';

import { memo } from 'react';

import type { TreeTransferItem, TreeTransferProps } from './types';

import InnerTreeTransfer from './InnerTreeTransfer';

const TreeTransfer = memo(InnerTreeTransfer) as (<RecordType extends TreeTransferItem = TreeTransferItem>(
  props: TreeTransferProps<RecordType>,
) => ReturnType<typeof InnerTreeTransfer<RecordType>>) &
  Pick<FC, 'displayName'>;

if (import.meta.env.DEV) {
  TreeTransfer.displayName = 'TreeTransfer';
}

export default TreeTransfer;

export type { TreeTransferItem, TreeTransferProps };
