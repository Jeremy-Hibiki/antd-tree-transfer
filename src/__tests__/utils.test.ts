import type { TreeDataNode } from 'antd';
import type { Key } from 'react';

import { generateLeftTree, generateRightTree, searchTreeFilter } from '../../src/utils';
import { TOTAL_NODES_COUNT, TREE_NODES, TreeNodeKey } from './fixtures/data';

function getNodeByKey(tree: TreeDataNode[], key: Key): TreeDataNode | undefined {
  if (tree.length === 0) {
    return undefined;
  }

  for (const node of tree) {
    if (node.key === key) {
      return node;
    }

    if (node.children) {
      const result = getNodeByKey(node.children, key);
      if (result) {
        return result;
      }
    }
  }

  return undefined;
}

function getNodesByStatus(tree: TreeDataNode[]) {
  return tree.reduce<Record<'disabled' | 'enabled', TreeDataNode[]>>(
    (acc, node) => {
      acc[node.disabled ? 'disabled' : 'enabled'].push(node);
      if (node.children) {
        const childNodes = getNodesByStatus(node.children);
        acc.disabled.push(...childNodes.disabled);
        acc.enabled.push(...childNodes.enabled);
      }
      return acc;
    },
    { disabled: [], enabled: [] },
  );
}

function notNil<T>(value: T | undefined): value is NonNullable<T> {
  return value !== undefined && value !== null;
}

describe('generate tree', () => {
  it('works', () => {
    const leftTree = generateLeftTree(TREE_NODES, []);
    expect(leftTree).toHaveLength(3);
    const rightTree = generateRightTree(TREE_NODES, []);
    expect(rightTree).toHaveLength(0);
  });

  describe('select none nodes', () => {
    it('left tree', () => {
      const tree = generateLeftTree(TREE_NODES, []);
      expect(tree).toMatchSnapshot();
    });

    it('right tree', () => {
      const tree = generateRightTree(TREE_NODES, []);
      expect(tree).toHaveLength(0);
    });
  });

  describe('select one of leaf nodes', () => {
    const selectedKeys = [TreeNodeKey.PARENT_3_CHILD_2];
    const leftTree = generateLeftTree(TREE_NODES, selectedKeys);
    const rightTree = generateRightTree(TREE_NODES, selectedKeys);

    it('left tree', () => {
      expect(leftTree).toMatchSnapshot();

      selectedKeys.forEach((key) => {
        const selectedNode = getNodeByKey(leftTree, key);
        expect(selectedNode?.disabled).toBeTruthy();
      });

      const parentNode = getNodeByKey(leftTree, TreeNodeKey.PARENT_3_WITH_MULTIPLE_CHILDREN);
      expect(parentNode?.disabled).toBeFalsy();

      const { disabled, enabled } = getNodesByStatus(leftTree);
      expect(disabled.length).toBe(1);
      expect(enabled.length).toBe(TOTAL_NODES_COUNT - 1);
    });

    it('right tree', () => {
      expect(rightTree).toHaveLength(1);
      expect(rightTree[0].children).toHaveLength(1);
      expect(rightTree).toMatchSnapshot();
    });
  });

  describe('select all of leaf nodes - 1', () => {
    const selectedKeys = [TreeNodeKey.PARENT_2_CHILD_1];
    const leftTree = generateLeftTree(TREE_NODES, selectedKeys);
    const rightTree = generateRightTree(TREE_NODES, selectedKeys);

    it('left tree', () => {
      expect(leftTree).toMatchSnapshot();

      selectedKeys.forEach((key) => {
        const selectedNode = getNodeByKey(leftTree, key);
        expect(selectedNode?.disabled).toBeTruthy();
      });

      const parentNode = getNodeByKey(leftTree, TreeNodeKey.PARENT_2_WITH_ONE_CHILD);
      expect(parentNode?.disabled).toBeTruthy();

      const { disabled, enabled } = getNodesByStatus(leftTree);
      expect(disabled.length).toBe(2);
      expect(enabled.length).toBe(TOTAL_NODES_COUNT - 2);
    });

    it('right tree', () => {
      expect(rightTree).toHaveLength(1);
      expect(rightTree[0].children).toHaveLength(1);
      expect(rightTree).toMatchSnapshot();
    });
  });

  describe('select all of leaf nodes - 2', () => {
    const selectedKeys = [TreeNodeKey.PARENT_3_CHILD_1, TreeNodeKey.PARENT_3_CHILD_2, TreeNodeKey.PARENT_3_CHILD_3];
    const leftTree = generateLeftTree(TREE_NODES, selectedKeys);
    const rightTree = generateRightTree(TREE_NODES, selectedKeys);

    it('left tree', () => {
      expect(leftTree).toMatchSnapshot();

      selectedKeys.forEach((key) => {
        const selectedNode = getNodeByKey(leftTree, key);
        expect(selectedNode?.disabled).toBeTruthy();
      });

      const parentNode = getNodeByKey(leftTree, TreeNodeKey.PARENT_3_WITH_MULTIPLE_CHILDREN);
      expect(parentNode?.disabled).toBeTruthy();

      const { disabled, enabled } = getNodesByStatus(leftTree);
      expect(disabled.length).toBe(4);
      expect(enabled.length).toBe(TOTAL_NODES_COUNT - 4);
    });

    it('right tree', () => {
      expect(rightTree).toHaveLength(1);
      expect(rightTree[0].children).toHaveLength(3);
      expect(rightTree).toMatchSnapshot();
    });
  });
});

describe('filter tree', () => {
  it('works', () => {
    const filteredItems = [getNodeByKey(TREE_NODES, TreeNodeKey.PARENT_3_CHILD_1)].filter(notNil);
    const filteredTree = searchTreeFilter(TREE_NODES, filteredItems);
    expect(filteredTree).toMatchSnapshot();
    expect(filteredTree).toHaveLength(1);
    expect(filteredTree[0].children).toHaveLength(1);
  });
});
