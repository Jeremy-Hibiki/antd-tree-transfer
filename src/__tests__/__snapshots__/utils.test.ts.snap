// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`filter tree > works 1`] = `
[
  {
    "children": [
      {
        "key": "parent-3-child-1",
        "title": "parent-3-child-1",
      },
    ],
    "key": "parent-3-with-multiple-children",
    "title": "parent-3-with-multiple-children",
  },
]
`;

exports[`generate tree > select all of leaf nodes - 1 > left tree 1`] = `
[
  {
    "disabled": false,
    "key": "parent-1-without-children",
    "title": "parent-1-without-children",
  },
  {
    "children": [
      {
        "disabled": true,
        "key": "parent-2-child-1",
        "title": "parent-2-child-1",
      },
    ],
    "disabled": true,
    "key": "parent-2-with-one-child",
    "title": "parent-2-with-one-child",
  },
  {
    "children": [
      {
        "disabled": false,
        "key": "parent-3-child-1",
        "title": "parent-3-child-1",
      },
      {
        "disabled": false,
        "key": "parent-3-child-2",
        "title": "parent-3-child-2",
      },
      {
        "disabled": false,
        "key": "parent-3-child-3",
        "title": "parent-3-child-3",
      },
    ],
    "disabled": false,
    "key": "parent-3-with-multiple-children",
    "title": "parent-3-with-multiple-children",
  },
]
`;

exports[`generate tree > select all of leaf nodes - 1 > right tree 1`] = `
[
  {
    "_keepNodeFlag": false,
    "children": [
      {
        "_keepNodeFlag": true,
        "children": [],
        "key": "parent-2-child-1",
        "title": "parent-2-child-1",
      },
    ],
    "key": "parent-2-with-one-child",
    "title": "parent-2-with-one-child",
  },
]
`;

exports[`generate tree > select all of leaf nodes - 2 > left tree 1`] = `
[
  {
    "disabled": false,
    "key": "parent-1-without-children",
    "title": "parent-1-without-children",
  },
  {
    "children": [
      {
        "disabled": false,
        "key": "parent-2-child-1",
        "title": "parent-2-child-1",
      },
    ],
    "disabled": false,
    "key": "parent-2-with-one-child",
    "title": "parent-2-with-one-child",
  },
  {
    "children": [
      {
        "disabled": true,
        "key": "parent-3-child-1",
        "title": "parent-3-child-1",
      },
      {
        "disabled": true,
        "key": "parent-3-child-2",
        "title": "parent-3-child-2",
      },
      {
        "disabled": true,
        "key": "parent-3-child-3",
        "title": "parent-3-child-3",
      },
    ],
    "disabled": true,
    "key": "parent-3-with-multiple-children",
    "title": "parent-3-with-multiple-children",
  },
]
`;

exports[`generate tree > select all of leaf nodes - 2 > right tree 1`] = `
[
  {
    "_keepNodeFlag": false,
    "children": [
      {
        "_keepNodeFlag": true,
        "children": [],
        "key": "parent-3-child-1",
        "title": "parent-3-child-1",
      },
      {
        "_keepNodeFlag": true,
        "children": [],
        "key": "parent-3-child-2",
        "title": "parent-3-child-2",
      },
      {
        "_keepNodeFlag": true,
        "children": [],
        "key": "parent-3-child-3",
        "title": "parent-3-child-3",
      },
    ],
    "key": "parent-3-with-multiple-children",
    "title": "parent-3-with-multiple-children",
  },
]
`;

exports[`generate tree > select none nodes > left tree 1`] = `
[
  {
    "disabled": false,
    "key": "parent-1-without-children",
    "title": "parent-1-without-children",
  },
  {
    "children": [
      {
        "disabled": false,
        "key": "parent-2-child-1",
        "title": "parent-2-child-1",
      },
    ],
    "disabled": false,
    "key": "parent-2-with-one-child",
    "title": "parent-2-with-one-child",
  },
  {
    "children": [
      {
        "disabled": false,
        "key": "parent-3-child-1",
        "title": "parent-3-child-1",
      },
      {
        "disabled": false,
        "key": "parent-3-child-2",
        "title": "parent-3-child-2",
      },
      {
        "disabled": false,
        "key": "parent-3-child-3",
        "title": "parent-3-child-3",
      },
    ],
    "disabled": false,
    "key": "parent-3-with-multiple-children",
    "title": "parent-3-with-multiple-children",
  },
]
`;

exports[`generate tree > select one of leaf nodes > left tree 1`] = `
[
  {
    "disabled": false,
    "key": "parent-1-without-children",
    "title": "parent-1-without-children",
  },
  {
    "children": [
      {
        "disabled": false,
        "key": "parent-2-child-1",
        "title": "parent-2-child-1",
      },
    ],
    "disabled": false,
    "key": "parent-2-with-one-child",
    "title": "parent-2-with-one-child",
  },
  {
    "children": [
      {
        "disabled": false,
        "key": "parent-3-child-1",
        "title": "parent-3-child-1",
      },
      {
        "disabled": true,
        "key": "parent-3-child-2",
        "title": "parent-3-child-2",
      },
      {
        "disabled": false,
        "key": "parent-3-child-3",
        "title": "parent-3-child-3",
      },
    ],
    "disabled": false,
    "key": "parent-3-with-multiple-children",
    "title": "parent-3-with-multiple-children",
  },
]
`;

exports[`generate tree > select one of leaf nodes > right tree 1`] = `
[
  {
    "_keepNodeFlag": false,
    "children": [
      {
        "_keepNodeFlag": true,
        "children": [],
        "key": "parent-3-child-2",
        "title": "parent-3-child-2",
      },
    ],
    "key": "parent-3-with-multiple-children",
    "title": "parent-3-with-multiple-children",
  },
]
`;
